File: packages/agent-cache/src/discovery.ts
import { hostname } from 'node:os';

import { AgentCacheUsageError } from './errors';
import type { Valkey } from './types';

export const PROTOCOL_VERSION = 1;

export const REGISTRY_KEY = '__betterdb:caches';
export const PROTOCOL_KEY = '__betterdb:protocol';
export const HEARTBEAT_KEY_PREFIX = '__betterdb:heartbeat:';

export const DEFAULT_HEARTBEAT_INTERVAL_MS = 30_000;
export const HEARTBEAT_TTL_SECONDS = 60;

export const TOOL_POLICIES_LIMIT = 500;

export const CACHE_TYPE = 'agent_cache' as const;
export type CacheType = typeof CACHE_TYPE;

export interface DiscoveryOptions {
  enabled?: boolean;
  heartbeatIntervalMs?: number;
  includeToolPolicies?: boolean;
}

export interface TierMarkerInfo {
  enabled: boolean;
  ttl_default?: number;
}

export interface MarkerMetadata {
  type: CacheType;
  prefix: string;
  version: string;
  protocol_version: number;
  capabilities: string[];
  stats_key: string;
  started_at: string;
  pid?: number;
  hostname?: string;
  [extra: string]: unknown;
}

export interface BuildAgentMetadataInput {
  name: string;
  version: string;
  tiers: {
    llm?: { ttl?: number };
    tool?: { ttl?: number };
    session?: { ttl?: number };
  };
  defaultTtl: number | undefined;
  toolPolicyNames: string[];
  hasCostTable: boolean;
  usesDefaultCostTable: boolean;
  startedAt: string;
  includeToolPolicies: boolean;
}

export function buildAgentMetadata(input: BuildAgentMetadataInput): MarkerMetadata {
  const tierMarker = (ttl: number | undefined): TierMarkerInfo => ({
    enabled: true,
    ttl_default: ttl ?? input.defaultTtl,
  });

  const metadata: MarkerMetadata = {
    type: 'agent_cache',
    prefix: input.name,
    version: input.version,
    protocol_version: PROTOCOL_VERSION,
    capabilities: ['tool_ttl_adjust', 'invalidate_by_tool', 'tool_effectiveness'],
    stats_key: `${input.name}:__stats`,
    tiers: {
      llm: tierMarker(input.tiers.llm?.ttl),
      tool: tierMarker(input.tiers.tool?.ttl),
      session: tierMarker(input.tiers.session?.ttl),
    },
    has_cost_table: input.hasCostTable,
    uses_default_cost_table: input.usesDefaultCostTable,
    started_at: input.startedAt,
    pid: process.pid,
    hostname: hostname(),
  };

  if (input.includeToolPolicies) {
    const names = input.toolPolicyNames;
    if (names.length > TOOL_POLICIES_LIMIT) {
      metadata.tool_policies = names.slice(0, TOOL_POLICIES_LIMIT);
      metadata.tool_policies_truncated = true;
    } else {
      metadata.tool_policies = [...names];
    }
  }

  return metadata;
}

export interface DiscoveryLogger {
  warn: (msg: string) => void;
  debug: (msg: string) => void;
}

const noopLogger: DiscoveryLogger = {
  warn: () => {},
  debug: () => {},
};

function errMsg(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

export interface DiscoveryManagerDeps {
  client: Valkey;
  name: string;
  buildMetadata: () => MarkerMetadata;
  heartbeatIntervalMs?: number;
  logger?: DiscoveryLogger;
  onWriteFailed?: () => void;
}

export class DiscoveryManager {
  private readonly client: Valkey;
  private readonly name: string;
  private readonly buildMetadata: () => MarkerMetadata;
  private readonly heartbeatIntervalMs: number;
  private readonly heartbeatKey: string;
  private readonly logger: DiscoveryLogger;
  private readonly onWriteFailed: () => void;

  private heartbeatHandle: ReturnType<typeof setInterval> | null = null;

  constructor(deps: DiscoveryManagerDeps) {
    this.client = deps.client;
    this.name = deps.name;
    this.buildMetadata = deps.buildMetadata;
    this.heartbeatIntervalMs = deps.heartbeatIntervalMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS;
    this.heartbeatKey = `${HEARTBEAT_KEY_PREFIX}${deps.name}`;
    this.logger = deps.logger ?? noopLogger;
    this.onWriteFailed = deps.onWriteFailed ?? (() => {});
  }

  async register(): Promise<void> {
    const existingJson = await this.safeHget();
    if (existingJson !== null) {
      this.checkCollision(existingJson);
    }

    await this.writeMetadata();
    await this.safeCall(
      () => this.client.set(PROTOCOL_KEY, String(PROTOCOL_VERSION), 'NX'),
      'SET protocol',
    );

    await this.writeHeartbeat();

    this.startHeartbeat();
  }

  async stop(opts: { deleteHeartbeat: boolean }): Promise<void> {
    if (this.heartbeatHandle) {
      clearInterval(this.heartbeatHandle);
      this.heartbeatHandle = null;
    }
    if (!opts.deleteHeartbeat) {
      return;
    }
    try {
      await this.client.del(this.heartbeatKey);
    } catch (err) {
      this.logger.debug(`discovery: DEL heartbeat failed: ${errMsg(err)}`);
    }
  }

  async tickHeartbeat(): Promise<void> {
    await this.writeHeartbeat();
    await this.writeMetadata();
    await this.safeCall(
      () => this.client.set(PROTOCOL_KEY, String(PROTOCOL_VERSION), 'NX'),
      'SET protocol (heartbeat)',
    );
  }

  private startHeartbeat(): void {
    const handle = setInterval(() => {
      void this.tickHeartbeat();
    }, this.heartbeatIntervalMs);
    handle.unref?.();
    this.heartbeatHandle = handle;
  }

  private async writeHeartbeat(): Promise<void> {
    const now = new Date().toISOString();
    try {
      await this.client.set(this.heartbeatKey, now, 'EX', HEARTBEAT_TTL_SECONDS);
    } catch (err) {
      this.logger.debug(`discovery: heartbeat SET failed: ${errMsg(err)}`);
      this.onWriteFailed();
    }
  }

  private async writeMetadata(): Promise<void> {
    let payload: string;
    try {
      payload = JSON.stringify(this.buildMetadata());
    } catch (err) {
      this.logger.warn(`discovery: metadata serialise failed: ${errMsg(err)}`);
      this.onWriteFailed();
      return;
    }
    await this.safeCall(() => this.client.hset(REGISTRY_KEY, this.name, payload), 'HSET registry');
  }

  private async safeHget(): Promise<string | null> {
    try {
      return await this.client.hget(REGISTRY_KEY, this.name);
    } catch (err) {
      this.logger.warn(`discovery: HGET registry failed: ${errMsg(err)}`);
      this.onWriteFailed();
      return null;
    }
  }

  private async safeCall(fn: () => Promise<unknown>, label: string): Promise<void> {
    try {
      await fn();
    } catch (err) {
      this.logger.warn(`discovery: ${label} failed: ${errMsg(err)}`);
      this.onWriteFailed();
    }
  }

  private checkCollision(existingJson: string): void {
    let parsed: Partial<MarkerMetadata>;
    try {
      parsed = JSON.parse(existingJson) as Partial<MarkerMetadata>;
    } catch {
      return;
    }
    if (parsed.type && parsed.type !== CACHE_TYPE) {
      throw new AgentCacheUsageError(
        `cache name collision: '${this.name}' is already registered as type '${String(parsed.type)}' on this Valkey instance`,
      );
    }
    const newMeta = this.buildMetadata();
    if (parsed.version && parsed.version !== newMeta.version) {
      this.logger.warn(
        `discovery: overwriting marker for '${this.name}' (existing version ${String(parsed.version)}, this version ${newMeta.version})`,
      );
    }
  }
}

File: packages/agent-cache-py/betterdb_agent_cache/discovery.py
from __future__ import annotations

import asyncio
import json
import os
import socket
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable

from .errors import AgentCacheUsageError
from .types import DiscoveryOptions  # single source of truth

PROTOCOL_VERSION = 1

REGISTRY_KEY = '__betterdb:caches'
PROTOCOL_KEY = '__betterdb:protocol'
HEARTBEAT_KEY_PREFIX = '__betterdb:heartbeat:'

DEFAULT_HEARTBEAT_INTERVAL_S = 30.0
HEARTBEAT_TTL_SECONDS = 60

TOOL_POLICIES_LIMIT = 500

CACHE_TYPE = 'agent_cache'


@dataclass
class TierMarkerInfo:
    enabled: bool
    ttl_default: int | None = None


@dataclass
class BuildAgentMetadataInput:
    name: str
    version: str
    tiers: dict[str, dict[str, Any]]
    default_ttl: int | None
    tool_policy_names: list[str]
    has_cost_table: bool
    uses_default_cost_table: bool
    started_at: str
    include_tool_policies: bool


def build_agent_metadata(input: BuildAgentMetadataInput) -> dict[str, Any]:
    def tier_marker(ttl: int | None) -> dict[str, Any]:
        effective_ttl = ttl if ttl is not None else input.default_ttl
        result: dict[str, Any] = {'enabled': True}
        if effective_ttl is not None:
            result['ttl_default'] = effective_ttl
        return result

    llm_ttl = input.tiers.get('llm', {}).get('ttl')
    tool_ttl = input.tiers.get('tool', {}).get('ttl')
    session_ttl = input.tiers.get('session', {}).get('ttl')

    metadata: dict[str, Any] = {
        'type': CACHE_TYPE,
        'prefix': input.name,
        'version': input.version,
        'protocol_version': PROTOCOL_VERSION,
        'capabilities': ['tool_ttl_adjust', 'invalidate_by_tool', 'tool_effectiveness'],
        'stats_key': f'{input.name}:__stats',
        'tiers': {
            'llm': tier_marker(llm_ttl),
            'tool': tier_marker(tool_ttl),
            'session': tier_marker(session_ttl),
        },
        'has_cost_table': input.has_cost_table,
        'uses_default_cost_table': input.uses_default_cost_table,
        'started_at': input.started_at,
        'pid': os.getpid(),
        'hostname': socket.gethostname(),
    }

    if input.include_tool_policies:
        names = input.tool_policy_names
        if len(names) > TOOL_POLICIES_LIMIT:
            metadata['tool_policies'] = names[:TOOL_POLICIES_LIMIT]
            metadata['tool_policies_truncated'] = True
        else:
            metadata['tool_policies'] = list(names)

    return metadata


def _err_msg(err: Exception) -> str:
    return str(err)


class DiscoveryManager:
    def __init__(
        self,
        *,
        client: Any,
        name: str,
        build_metadata: Callable[[], dict[str, Any]],
        heartbeat_interval_s: float = DEFAULT_HEARTBEAT_INTERVAL_S,
        logger: Any = None,
        on_write_failed: Callable[[], None] | None = None,
    ) -> None:
        self._client = client
        self._name = name
        self._build_metadata = build_metadata
        self._heartbeat_interval_s = heartbeat_interval_s
        self._heartbeat_key = f'{HEARTBEAT_KEY_PREFIX}{name}'
        self._logger = logger
        self._on_write_failed: Callable[[], None] = on_write_failed or (lambda: None)
        self._heartbeat_task: asyncio.Task[None] | None = None

    def _warn(self, msg: str) -> None:
        if self._logger is not None:
            self._logger.warning(msg)

    def _debug(self, msg: str) -> None:
        if self._logger is not None:
            self._logger.debug(msg)

    async def register(self) -> None:
        existing_json = await self._safe_hget()
        if existing_json is not None:
            self._check_collision(existing_json)

        await self._write_metadata()
        await self._safe_call(
            lambda: self._client.set(PROTOCOL_KEY, str(PROTOCOL_VERSION), 'NX'),
            'SET protocol',
        )

        await self._write_heartbeat()
        self._start_heartbeat()

    async def stop(self, *, delete_heartbeat: bool) -> None:
        if self._heartbeat_task is not None:
            self._heartbeat_task.cancel()
            try:
                await self._heartbeat_task
            except (asyncio.CancelledError, Exception):
                pass
            self._heartbeat_task = None

        if not delete_heartbeat:
            return

        try:
            await self._client.delete(self._heartbeat_key)
        except Exception as err:
            self._debug(f'discovery: DEL heartbeat failed: {_err_msg(err)}')

    async def tick_heartbeat(self) -> None:
        await self._write_heartbeat()
        await self._write_metadata()
        await self._safe_call(
            lambda: self._client.set(PROTOCOL_KEY, str(PROTOCOL_VERSION), 'NX'),
            'SET protocol (heartbeat)',
        )

    def _start_heartbeat(self) -> None:
        async def _loop() -> None:
            try:
                while True:
                    await asyncio.sleep(self._heartbeat_interval_s)
                    await self.tick_heartbeat()
            except asyncio.CancelledError:
                pass

        self._heartbeat_task = asyncio.create_task(_loop())

    async def _write_heartbeat(self) -> None:
        now = datetime.now(timezone.utc).isoformat()
        try:
            await self._client.set(self._heartbeat_key, now, 'EX', HEARTBEAT_TTL_SECONDS)
        except Exception as err:
            self._debug(f'discovery: heartbeat SET failed: {_err_msg(err)}')
            self._on_write_failed()

    async def _write_metadata(self) -> None:
        try:
            payload = json.dumps(self._build_metadata())
        except Exception as err:
            self._warn(f'discovery: metadata serialise failed: {_err_msg(err)}')
            self._on_write_failed()
            return
        await self._safe_call(
            lambda: self._client.hset(REGISTRY_KEY, self._name, payload),
            'HSET registry',
        )

    async def _safe_hget(self) -> str | None:
        try:
            result = await self._client.hget(REGISTRY_KEY, self._name)
            if result is None:
                return None
            return result.decode() if isinstance(result, bytes) else result
        except Exception as err:
            self._warn(f'discovery: HGET registry failed: {_err_msg(err)}')
            self._on_write_failed()
            return None

    async def _safe_call(self, fn: Callable[[], Any], label: str) -> None:
        try:
            await fn()
        except Exception as err:
            self._warn(f'discovery: {label} failed: {_err_msg(err)}')
            self._on_write_failed()

    def _check_collision(self, existing_json: str) -> None:
        try:
            parsed: dict[str, Any] = json.loads(existing_json)
        except Exception:
            return

        existing_type = parsed.get('type')
        if existing_type and existing_type != CACHE_TYPE:
            raise AgentCacheUsageError(
                f"cache name collision: '{self._name}' is already registered as type "
                f"'{existing_type}' on this Valkey instance"
            )

        new_meta = self._build_metadata()
        existing_version = parsed.get('version')
        new_version = new_meta.get('version')
        if existing_version and existing_version != new_version:
            self._warn(
                f"discovery: overwriting marker for '{self._name}' "
                f"(existing version {existing_version}, this version {new_version})"
            )
